Ensure that Random classes successfully produce the requested bytes

C0deH4cker opened the pull request:
The biggest issue solved by this PR is that `URandom` incorrectly reads from /dev/urandom as string data. /dev/urandom contains binary data that must not be treated as a string. Currently, bytes are being read from this device using `fgets()`. [`man 3 fgets`](https://linux.die.net/man/3/fgets) says the following:

> Reading stops after an **EOF** or a newline.

Because of this, if one of the bytes from /dev/urandom happens to be a newline (maybe `\0` as well), then `fgets()` will stop reading. That means the rest of the bytes from this point on will be all zeros.

This is evident by visual inspection of the output from `URandom()` as well. For example, I'm using `URandom` to generate secure tokens for use as session tokens among other things. I looked at the tokens stored and found the following: `WPxXr5UKAAAAAAAAAAAAAA`. The `A`s at the end are the base64 representation of zero bytes. In addition to this, every single token in the database ends with one of 4 characters: `Q`, `g`, `w`, or `A`. The tokens I am generating are only 16 bytes of randomness, but the issue is even more evident for larger chunks of random data, since every byte read has a 1/256 (1/128 if `\0` also makes `fgets()` stop reading) chance of making every byte after that be zero.

In addition to fixing this issue by changing from `fgets()` to `fread()` (which _can_ be used to read binary data), the return code of all random data reading functions is checked to ensure they successfully returned the requested number of securely random bytes. For the `CryptoRandom` and `PseudoRandom` classes, I read [`man 3 RAND_bytes`](https://linux.die.net/man/3/rand_bytes) to find that these functions return 1 on success.

tanner0101 commented on Sources/Random/CryptoRandom.swift:
```diff
@@ -6,7 +6,11 @@ public final class CryptoRandom: Random {
+		guard RAND_bytes(&random, Int32(num)) == 1 else {
+			// If the requested number of random bytes couldn't be read,
+			// we need to fail fast and hard.
+			abort()
```
We should throw an error instead of `abort()`. If someone is using this in vapor and doesn't have something like supervisor setup, this will kill their web service.

LoganWright replied:
I think the issue here is that to throw properly, the protocol has to be edited, which would be a breaking change. I think in security situations, these types of things are more acceptable. 

I'm more for figuring out a throw or some other option, possibly trying a few times, then logging. Anything that won't crash is on the table.

@C0deH4cker fill in if I'm missing something.

tanner0101 replied:
Just saw the messages in #development. I can't believe the `bytes` method doesn't throw... >_<.

Can we move the protocol to throwing and add a non-throwing extension that gets a deprecation warning and has a `try!` ?

C0deH4cker replied:
Ah yeah I can make this change. I'll update the PR!

C0deH4cker commented:
@tanner0101: I discussed this in Slack briefly. This method conforms to a protocol not marked as `throws` and not able to return some other indicator of failure. What do you recommend to be done in this case? Should I change the protocol to be `throws`? This is a breaking change, hence why I avoided it. An alternative possibility is we could log an error message before aborting, not sure if that satisfies your concerns though. Let me know what you suggest and I'll update the PR accordingly!

C0deH4cker commented:
I added a commit to hopefully address your concerns. I deprecated the old non-throwing bytes method and the properties that used it and replaced them all with throwing variants. This is a breaking change but only where users have implemented their own classes that conform to `Random`. All other code should continue working as-is but with deprecation warnings.

C0deH4cker commented:
Oops, `BCryptSalt` uses the now-deprecated version of `bytes()`

C0deH4cker commented:
Okay, should be all set now @tanner0101 @LoganWright!

tanner0101 commented:
Moved to #20
